tests/nvmeof/workflows/ha.py

```diff
@@ -0,0 +1,421 @@
+            if _state["anagrp-id"] == ana_id:
+                if _state["Availability"] == state:
+                    return True
+                return False
+        return False
+
+    def get_optimized_state(self, failed_ana_id):
+        """Fetch the Optimized ANA states for failed gateway.
+
+        Args:
+            gateway: The gateway which is operational.
+            failed_ana_id: failed gateway ANA Group Id.
+
+        Returns:
+            gateways which shows ACTIVE state for failed ANA Group Id
+        """
+        # get ANA states
+        ana_states = self.ana_states()
+
+        # Fetch failed ANA Group Id in ACTIVE state
+        found = []
+
+        for ana_gw_id, state in ana_states.items():
+            if (
+                state["Availability"] == "AVAILABLE"
+                and state.get(failed_ana_id) == "ACTIVE"
+            ):
+                found.append({ana_gw_id: state})
+
+        return found
+
+    def failover(self, gateway, fail_tool="systemctl"):
+        """HA Failover on the NVMeoF Gateways.
+
+        Initiate Failover
+        - List the gateways which not have failures.
+        - Initiate failures on GWs which has to be down systemctl/daemon stop
+        - Validate the ANA states of failed GWs are optimized in one of the other working GWs.
+
+        Post Failover Validation
+        - List out namespaces associated with the failed Gateways using ANA group ids.
+        - Check for 5 Consecutive times for the increments in write/read to validate IO continuation.
+        """
+        impending_fail_gws, _ = self.catogorize(gateway)
+
+        LOG.info(
+            f"Gateways to-be failed : {[i.node.hostname for i in impending_fail_gws]}"
+        )
+
+        # Initiate Failover
+        with parallel() as p:
+            for gw in impending_fail_gws:
+                LOG.info(
+                    f"[{gw.node.hostname}]: Failing NVMe Service Systemctl command"
+                )
+                service = gw.system_unit_id
+                p.spawn(gw.systemctl.stop, service)
+
+                for w in WaitUntil():
+                    if (
+                        not gw.systemctl.is_active(service)
+                    ) and self.check_gateway_availability(
+                        gw.ana_group_id, state="UNAVAILABLE"
+                    ):
+                        LOG.info(f"[{service}] NVMeofGW service Stopped successfully.")
+                        break
+                    LOG.warning(
+                        f"[{service}] Stopping NVMeofGW service Failed. try again !!!!"
+                    )
+
+            if w.expired:
+                raise TimeoutError(
+                    f"{service} Stopping NVMeofGW service failed after 60s timeout.."
+                )
+
+        # Check for Active or Optimized path for failed GWs.
+        # Validate that there should be only one active path.
+        total_outage_time = None
+        fail_time = datetime.datetime.now()
```

Comment:
In a correct way, I think, right fail_time is when IO pauses or for time being I think we need to move this line to just after 

p.spawn(gw.systemctl.stop, service)

Reply:
Considering this next PR. where we will get outage time per gateway.